plutus-chain-index/README_alpha-release_Marconi.md

```diff
@@ -0,0 +1,75 @@
+
+### Indexing solution ### 
+
+Any indexing solution must deal with blockchain rollbacks, which can happen as the network achieves consensus. Marconi is designed to deal with rollbacks as elegantly as possible while simultaneously finding an effective balance between memory, disk and CPU usage. 
+
+Managing rollbacks is very simple and fast. Marconi drops the events that were rolled back. No complicated logic is required to undo the projection of the list of events on disk, which Marconi would need if it stored everything on disk as quickly as possible. 
+
+Making `K` configurable makes the design quite scalable. Developers do not usually need to guard themselves against rollbacks by `K` blocks, so they can choose to store only 10 events in memory, for example, thereby sacrificing the speed by which Marconi handles rollbacks (since rollbacks on data that is persisted on disk is more expensive than data that is in memory). 
+
+In the event of a restart, recovery is very simple. Marconi stores only fully confirmed transactions, so there is nothing to do other than to load the state persisted on disk and to follow the blockchain from the corresponding chain-point. 
+
+## Queries
+
+The indexed data is accessible through queries. There are no constraints on the format of queries or on the results. Both are identified by a type variable that the indexer exposes. The user, such as a dApp developer, provides the code for both. 
+
+This is currently the only area where you can customize the indexer by providing functions that handle queries and storage. Note that the queries have to run on both the in-memory data and the data that is stored.
+
+# Installation
+
+Prerequisite: You already have a Cardano node running. 
+
+## Example command for launching Cardano node 
+
+```
+cardano-node run \
+    --config mainnet-config.json \
+    --topology mainnet-topology.json \
+    --database-path db \
+    --socket-path /tmp/node-mainnet.sock \
+    --port 3003
+```
+
+## Marconi executable
+
+The `marconi` executable is available as a nix flake.
+
+You may either clone the [`plutus-apps`](https://github.com/input-output-hk/plutus-apps)
+repository and run from the top-level:
+``` 
+nix build .#marconi
+``` 
+Or you may run from anywhere:
+```
+nix build github:input-output-hk/plutus-apps#marconi
+``` 
+
+Both commands will produce a `result` directory containing the executable 
+`result/bin/marconi`.
+
+## Example command for launching Marconi
+
+```
+./result/bin/marconi --socket-path /tmp/node-mainnet.sock --mainnet --datum-db datum.sqlite3 --utxo-db utxo.sqlite3
+```
```

Comment:
```suggestion
```
## Querying the SQL Database

A. Datum indexer

The database contains only one table defined as:

```
CREATE TABLE kv_datumhsh_datum (datumHash TEXT PRIMARY KEY, datum BLOB, slotNo INT);
```

The basic query that you can run is:

```
SELECT datum from kv_datumhsh_datum WHERE datumHash = ?;
```

B. Utxo indexer

The database contains two tables, one for unspent transactions and one for spent inputs. They are defined as follows:

```
CREATE TABLE utxos (address TEXT NOT NULL, txId TEXT NOT NULL, inputIx INT NOT NULL);
CREATE TABLE spent (txId TEXT NOT NULL, inputIx INT NOT NULL);
```

After you are fully synchronised you may run a couple of SQL statements to make sure that the query speed is optimised.

First, to create indices:

```
CREATE INDEX utxo_address ON utxos (address);
CREATE INDEX utxo_refs ON utxos (txId, inputIx);
CREATE INDEX spent_refs ON spent (txId, inputIx);
```

Then, to match inputs and outputs and trim the database:

```
DELETE FROM utxos WHERE utxos.rowid IN (SELECT utxos.rowid FROM utxos LEFT JOIN spent on utxos.txId = spent.txId AND utxos.inputIx = spent.inputIx WHERE spent.txId IS NOT NULL);
VACUUM;
```

And we can run the query using:

```
SELECT address, txId, inputIx FROM utxos LEFT JOIN spent ON utxos.txId = spent.txId AND utxos.inputIx = spent.inputIx WHERE utxos.txId IS NULL AND utxos.address = ?
```